Tracker/Handler.php

```diff
@@ -42,6 +42,11 @@ public function process(Tracker $tracker, RequestSet $requestSet)
         $queueManager->addRequestSetToQueues($requestSet);
         $tracker->setCountOfLoggedRequests($requestSet->getNumberOfRequests());
 
+        $requests = $requestSet->getRequests();
+        foreach ($requests as $request) {
+            $request->setThirdPartyCookie($request->getVisitorId());
```

Comment:
This looks good to me. FYI: I reckon this won't work as expected for a bulk request with different visitor ids.

If there were multiple requests through a bulk request, and the bulk request contains requests for visitorIds, then it would likely always overwrite the same cookie `name=0` with the last request of the request set in `setThirdPartyCookie` where it executes `$cookie->set(0, bin2hex($idVisitor));`.

Bulk requests with different visitorIds should be only the case for the log importer or when sending maybe lots of requests through the Piwik PHP tracker (or Java tracker, Android, iOS, ...) but in these cases the third party cookie wouldn't be used anyway I presume. 

So for the regular JS tracker this shouldn't be a problem and neither for other cases.

Reply:
Interesting point. What about the other direction, when server logs are imported with the log importert tool. I took a quick look at its source code, and could not find any code that forwards the third party cookie from the logs to the tracker, so I guess it is not supported right now. Does this mean, that in this case we lose the whole third party functionality, and basically "corrupt" the data by using the log importer? Would it maybe be a good idea to create a patch for the log importer, so that it also reads the third party cookie from the logs, and only creates bulk requests for log entries with the same third party cookie?

Reply:
In the case where the browser does not already have a third party cookie and Piwik is turned off, would it maybe be a good idea, to just generate a random new visitor id and send it to the browser? Or would this cookie then be rejected by Piwik later on, because the visitor is unknown?

Reply:
@mattab I think you can answer this better. I think eg in the apache logs there is no information about cookies so they can't be replayed / applied. 

>  would it maybe be a good idea, to just generate a random new visitor id and send it to the browser

not quite sure what you mean by Piwik is turned off?

Reply:
nginx for example can log cookies ( if I did not misunderstand the documentation). With "turned off" I mean the case when Piwik does not track anything and all requests go to the webserver log (apache/ngingx).

Reply:
I don't really know what ngnix can log but I would presume it can. And then it should be possible to forward the third party cookie when importing logs. Not sure if there would be any advantage unless the cookie really already exists in Piwik. I presume it would be actually useful eg during a (short) maintenance mode where maybe many users would already have a cookie, and then this cookie could be considered but it still wouldn't be sent back to the user. And possible the idVisitor would already exist in the tracking URL anyway? Not sure! @mattab ?

When Piwik is turned off, it wouldn't be possible to send the visitor cookie I suppose, unless I misunderstand something.